src/transport/raw/TCP.cpp

```diff
@@ -349,16 +351,45 @@ CHIP_ERROR TCPBase::ProcessReceivedBuffer(Inet::TCPEndPoint * endPoint, const Pe
             buffer->ConsumeHead(messageSize);
             SuccessOrExit(err);
 
-            err = endPoint->AckReceive(messageSize);
+            err = endPoint->AckReceive(payloadSize);
             SuccessOrExit(err);
             continue;
         }
 
-        // Buffer is incomplete if we reach this point
+        // If we reach this point, the head buffer does not contain a complete message.
+        // We may or may not have a complete message in a buffer chain.
         if (buffer->HasChainedBuffer())
         {
-            buffer->CompactHead();
-            continue;
+            if (payloadSize > System::PacketBuffer::kMaxSizeWithoutReserve)
+            {
+                // This message will not fit in a maximum-size buffer.
+                if (payloadSize <= buffer->TotalLength())
+                {
+                    // We have the oversize message in the current buffer chain.
+                    // Since it's oversize, it is definitely not a valid CHIP message, so we angrily toss it aside.
+                    buffer.Consume(payloadSize);
+                    err = endPoint->AckReceive(payloadSize);
+                    SuccessOrExit(err);
+                    continue;
+                }
+                // If we don't yet have the complete message, proceed along to open the receive window.
+            }
+            else
+            {
+                if (payloadSize > buffer->MaxDataLength())
+                {
+                    // The current buffer head is too small for the message. We're gonna need a bigger buf.
+                    System::PacketBufferHandle newHead = System::PacketBufferHandle::NewWithData(
+                        buffer->Start(), buffer->DataLength(), static_cast<uint16_t>(payloadSize - buffer->DataLength()),
+                        0 /* reserved size */);
+                    VerifyOrExit(!newHead.IsNull(), err = CHIP_ERROR_NO_MEMORY);
```

Comment:
So we should probably document that this is going to push the buffer back into the queue and wait to see if we can get more memory later?  Or should we consume the message and throw it away if we can't allocate a buffer to handle it with?

Reply:
I don't think we have a good low-memory story yet in general. It would be better for the lowest-level interface to drop packets that arrive faster than we can handle them.

Reply:
Well, as it stands, the only way the current approach can work is if back-pressure is applied by re-queuing and _not_ TCP acking.  That does seem like a way to stall communication forever.

If we want to discard a message mid-payload, we need a new sort of state machine to track the fact that we are in the middle of receiving a message, but we've discarded its contents.